clang-11 static analysis at 1e280ca517976f464116a460729cd0a34b5dd682

```
$ clang-11  --version
clang version 11.0.0-++20200523063809+304b0ed4039-1~exp1~20200523164413.1701
$ scan-build-11 cmake .
make -j40 2> clang-analyze.log
```

output:
```
In file included from /home/gdevenyi/scratch/zlib-ng/arch/x86/deflate_quick.c:29:
/home/gdevenyi/scratch/zlib-ng/arch/x86/../../trees_emit.h:75:13: warning: ‘bi_flush’ defined but not used [-Wunused-function]
   75 | static void bi_flush(deflate_state *s) {
      |             ^~~~~~~~
In file included from /home/gdevenyi/scratch/zlib-ng/arch/x86/deflate_quick.c:29:
/home/gdevenyi/scratch/zlib-ng/arch/x86/../../trees_emit.h:75:13: warning: ‘bi_flush’ defined but not used [-Wunused-function]
   75 | static void bi_flush(deflate_state *s) {
      |             ^~~~~~~~
/home/gdevenyi/scratch/zlib-ng/deflate_slow.c:138:9: warning: Value stored to 'bflush' is never read [deadcode.DeadStores]
        bflush = zng_tr_tally_lit(s, s->window[s->strstart-1]);
        ^        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
/home/gdevenyi/scratch/zlib-ng/deflate_slow.c:138:9: warning: Value stored to 'bflush' is never read [deadcode.DeadStores]
        bflush = zng_tr_tally_lit(s, s->window[s->strstart-1]);
        ^        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
/home/gdevenyi/scratch/zlib-ng/test/example.c: In function ‘test_gzio’:
/home/gdevenyi/scratch/zlib-ng/test/example.c:199:16: warning: pointer targets in passing argument 1 of ‘strcmp’ differ in signedness [-Wpointer-sign]
  199 |     if (strcmp(uncompr, hello) != 0) {
      |                ^~~~~~~
      |                |
      |                unsigned char *
In file included from /home/gdevenyi/scratch/zlib-ng/zutil.h:22,
                 from /home/gdevenyi/scratch/zlib-ng/deflate.h:13,
                 from /home/gdevenyi/scratch/zlib-ng/test/example.c:12:
/usr/include/string.h:136:32: note: expected ‘const char *’ but argument is of type ‘unsigned char *’
  136 | extern int strcmp (const char *__s1, const char *__s2)
      |                    ~~~~~~~~~~~~^~~~
/home/gdevenyi/scratch/zlib-ng/test/example.c:84:14: warning: variable ‘read’ set but not used [-Wunused-but-set-variable]
   84 |     int err, read;
      |              ^~~~
/home/gdevenyi/scratch/zlib-ng/test/example.c: In function ‘test_gzio’:
/home/gdevenyi/scratch/zlib-ng/test/example.c:199:16: warning: pointer targets in passing argument 1 of ‘strcmp’ differ in signedness [-Wpointer-sign]
  199 |     if (strcmp(uncompr, hello) != 0) {
      |                ^~~~~~~
      |                |
      |                unsigned char *
In file included from /home/gdevenyi/scratch/zlib-ng/zutil.h:22,
                 from /home/gdevenyi/scratch/zlib-ng/deflate.h:13,
                 from /home/gdevenyi/scratch/zlib-ng/test/example.c:12:
/usr/include/string.h:136:32: note: expected ‘const char *’ but argument is of type ‘unsigned char *’
  136 | extern int strcmp (const char *__s1, const char *__s2)
      |                    ~~~~~~~~~~~~^~~~
/home/gdevenyi/scratch/zlib-ng/test/example.c:84:14: warning: variable ‘read’ set but not used [-Wunused-but-set-variable]
   84 |     int err, read;
      |              ^~~~
/home/gdevenyi/scratch/zlib-ng/test/example.c:198:5: warning: Value stored to 'read' is never read [deadcode.DeadStores]
    read = PREFIX(gzfread)(uncompr, uncomprLen, 1, file);
    ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
/home/gdevenyi/scratch/zlib-ng/test/example.c:198:5: warning: Value stored to 'read' is never read [deadcode.DeadStores]
    read = PREFIX(gzfread)(uncompr, uncomprLen, 1, file);
    ^      ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 warning generated.
/home/gdevenyi/scratch/zlib-ng/test/infcover.c:268:5: warning: Use of memory after it is freed [unix.Malloc]
    return re == NULL ? in : re;
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/gdevenyi/scratch/zlib-ng/test/infcover.c:466:5: warning: Value stored to 'buf' is never read [deadcode.DeadStores]
    buf += len;
    ^      ~~~
2 warnings generated.

```

This static analysis warning appears to be wrong: 
```
infcover.c:268:5: warning: Use of memory after it is freed [unix.Malloc]
    return re == NULL ? in : re;
```
Is there another way to write this to disable the warning?

--- a/test/infcover.c
+++ b/test/infcover.c
@@ -242,8 +242,11 @@ static void mem_done(PREFIX3(stream) *strm, char *prefix) {
 static unsigned char *h2b(const char *hex, unsigned *len) {
     unsigned char *in, *re;
     unsigned next, val;
+    unsigned inlen;
 
-    in = malloc((strlen(hex) + 1) >> 1);
+    inlen = (strlen(hex) + 1) >> 1;
+    assert(inlen != 0);     /* tell static analyzer we won't call malloc(0) */
+    in = malloc(inlen);
     if (in == NULL)
         return NULL;
     next = 0;
@@ -264,6 +267,7 @@ static unsigned char *h2b(const char *hex, unsigned *len) {
     } while (*hex++);       /* go through the loop with the terminating null */
     if (len != NULL)
         *len = next;
+    assert(next != 0);      /* tell static analyzer we won't call realloc(in, 0) */
     re = realloc(in, next);
     return re == NULL ? in : re;
 }

